nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,573 @@
+      carrot_arc_pub_->publish(arc_pts_msg);
+      return true;
+    }
+  }
+
+  carrot_arc_pub_->publish(arc_pts_msg);
+
+  return false;
+}
+
+bool RegulatedPurePursuitController::inCollision(const double & x, const double & y)
+{
+  unsigned int mx, my;
+  costmap_->worldToMap(x, y, mx, my);
+
+  unsigned char cost = costmap_->getCost(mx, my);
+
+  if (costmap_ros_->getLayeredCostmap()->isTrackingUnknown()) {
+    return cost >= INSCRIBED_INFLATED_OBSTACLE && cost != NO_INFORMATION;
+  } else {
+    return cost >= INSCRIBED_INFLATED_OBSTACLE;
+  }
+}
+
+double RegulatedPurePursuitController::costAtPose(const double & x, const double & y)
+{
+  unsigned int mx, my;
+  costmap_->worldToMap(x, y, mx, my);
+
+  unsigned char cost = costmap_->getCost(mx, my);
+  return static_cast<double>(cost);
+}
+
+void RegulatedPurePursuitController::applyConstraints(
+  double & linear_vel,
+  const double & dist_error, const double & lookahead_dist,
+  const double & curvature, const geometry_msgs::msg::Twist & curr_speed,
+  const double & pose_cost)
+{
+  // limit the linear velocity by curvature
+  const double radius = fabs(1.0 / curvature);
+  const double & min_rad = regulated_linear_scaling_min_radius_;
+  if (use_regulated_linear_velocity_scaling_ && radius < min_rad) {
+    linear_vel *= 1.0 - (fabs(radius - min_rad) / min_rad);
+    if (linear_vel < regulated_linear_scaling_min_speed_) {
+      linear_vel = regulated_linear_scaling_min_speed_;
+    }
+  }
+
+  // limit the linear velocity by proximity to obstacles
+  if (use_cost_regulated_linear_velocity_scaling_ &&
+    pose_cost != static_cast<double>(NO_INFORMATION))
+  {
+    // Note(stevemacenski): We can use the cost at a pose as a derived value proportional to
+    // distance to obstacle since we lack a distance map. [0-128] is the freespace costed
+    // range, above 128 is possibly inscribed, so it should max out at 128 to be a minimum speed
+    // in when in questionable states. This creates a linear mapping of
+    // cost [0, 128] to speed [min regulated speed, linear vel]
+    double max_non_collision = 128.0;
+    if (pose_cost > max_non_collision) {
+      linear_vel = regulated_linear_scaling_min_speed_;
+    } else {
+      const double slope = (regulated_linear_scaling_min_speed_ - linear_vel) / max_non_collision;
+      linear_vel = slope * pose_cost + linear_vel;
+    }
+  }
+
+  // if the actual lookahead distance is shorter than requested, that means we're at the
+  // end of the path. We'll scale linear velocity by error to slow to a smooth stop
+  if (use_approach_vel_scaling_ && dist_error > 2.0 * costmap_->getResolution()) {
+    double velocity_scaling = 1.0 - (dist_error / lookahead_dist);
+    double unbounded_vel = linear_vel * velocity_scaling;
+    if (unbounded_vel < min_approach_linear_velocity_) {
+      linear_vel = min_approach_linear_velocity_;
+    } else {
+      linear_vel = linear_vel * velocity_scaling;
```

Comment:
We want them all to count against, that's why we have a minimum set velocity so that it doesn't dip below something reasonable with each of these cost functions applied (and kinematic checking shown below to make sure any rapid changes are still feasible). If we're in a high curvature **and** high cost space, wouldn't you want to be the conservative-ness of both these heuristics? I agree the multiplication of each is a little aggressive. Maybe this is something we can play with before merging this into Nav2 main but for the tutorial stuff here I think this is OK. 

Gaming this out: 
- On normal paths not near the goal, the only 2 that can be applied are the curvature and cost. Curvature is a multiplier on the base max linear speed and the cost is a negative linear function to reduce it. So there isn't compounding multiplier here. 
- On approaching the goal, the other functions basically don't matter. The curvature will naturally be very low since we're driving straight into it. The cost might impact things, but the closer we get, the more likely that the velocity scaling on approach cost function will take the bulk of the work. 

It might be good to do either-or on the cost / curvature, but there's a `regulated_linear_scaling_min_speed_` min set so that it can never get _too_ low. I'll add a TODO to test this and play around before a PR is submitted into the main repo

```
  double & dt = control_duration_;
  const double max_feasible_linear_speed = curr_speed.linear.x + max_linear_accel_ * dt;
  const double min_feasible_linear_speed = curr_speed.linear.x - max_linear_decel_ * dt;
  linear_vel = std::clamp(linear_vel, min_feasible_linear_speed, max_feasible_linear_speed);
```

Reply:
@shrijitsingh99 here's a patch for doing the max of the heuristics rather than just the compounding of them https://drive.google.com/file/d/1MqOrdVz-ZQcqbjoJ_CaXACdIPJVJ2LWN/view?usp=sharing you can apply it to this branch with `git apply <path name>.patch`. 

Let me know what you think. It's better-ish but hard to really tell in the turtlebot3 sandbox world